src/main/java/com/gotocompany/firehose/sink/grpc/client/GrpcClient.java

```diff
@@ -41,39 +45,47 @@ public GrpcClient(FirehoseInstrumentation firehoseInstrumentation, GrpcSinkConfi
         this.managedChannel = managedChannel;
     }
 
-    public DynamicMessage execute(byte[] logMessage, Headers headers) {
-
+    public void initialize() {
         MethodDescriptor.Marshaller<byte[]> marshaller = getMarshaller();
-        DynamicMessage dynamicMessage;
-
-        try {
+        this.methodDescriptor = MethodDescriptor.newBuilder(marshaller, marshaller)
+                .setType(MethodDescriptor.MethodType.UNARY)
+                .setFullMethodName(grpcSinkConfig.getSinkGrpcMethodUrl())
+                .build();
+    }
 
+    public DynamicMessage execute(byte[] logMessage, Headers headers) {
 
+        try {
             Metadata metadata = new Metadata();
             for (Header header : headers) {
                 metadata.put(Metadata.Key.of(header.key(), Metadata.ASCII_STRING_MARSHALLER), new String(header.value()));
             }
-
             Channel decoratedChannel = ClientInterceptors.intercept(managedChannel,
                      MetadataUtils.newAttachHeadersInterceptor(metadata));
             byte[] response = ClientCalls.blockingUnaryCall(
                     decoratedChannel,
-                    MethodDescriptor.newBuilder(marshaller, marshaller)
-                            .setType(MethodDescriptor.MethodType.UNARY)
-                            .setFullMethodName(grpcSinkConfig.getSinkGrpcMethodUrl())
-                            .build(),
-                    CallOptions.DEFAULT,
+                    methodDescriptor,
+                    decorateCallOptions(CallOptions.DEFAULT),
                     logMessage);
 
-            dynamicMessage = stencilClient.parse(grpcSinkConfig.getSinkGrpcResponseSchemaProtoClass(), response);
+            return stencilClient.parse(grpcSinkConfig.getSinkGrpcResponseSchemaProtoClass(), response);
 
+        } catch (StatusRuntimeException sre) {
+            firehoseInstrumentation.logWarn(sre.getMessage());
+            firehoseInstrumentation.incrementCounter(Metrics.SINK_GRPC_ERROR_TOTAL,  "status=" + sre.getStatus().getCode());
         } catch (Exception e) {
+            e.printStackTrace();
             firehoseInstrumentation.logWarn(e.getMessage());
-            dynamicMessage = DynamicMessage.newBuilder(this.stencilClient.get(this.grpcSinkConfig.getSinkGrpcResponseSchemaProtoClass())).build();
-
+            firehoseInstrumentation.incrementCounter(Metrics.SINK_GRPC_ERROR_TOTAL, "status=UNIDENTIFIED");
         }
+        return DynamicMessage.newBuilder(this.stencilClient.get(this.grpcSinkConfig.getSinkGrpcResponseSchemaProtoClass())).build();
+    }
 
-        return dynamicMessage;
+    protected CallOptions decorateCallOptions(CallOptions defaultCallOption) {
```

Comment:
`    protected CallOptions decorateCallOptions() {
        CallOptions defaultCallOption = CallOptions.DEFAULT;
        if (grpcSinkConfig.getSinkGrpcArgDeadlineMS() != null && grpcSinkConfig.getSinkGrpcArgDeadlineMS() > 0) {
            return defaultCallOption.withDeadlineAfter(grpcSinkConfig.getSinkGrpcArgDeadlineMS(), TimeUnit.MILLISECONDS);
        }
        return defaultCallOption;
    }`

Reply:
This needs to be checked.. What to set deadline when the default or too large value of keepalive is set.. Maybe can take another config for this?

Reply:
Default grpc client don't set any default value. Not sure what would be right default value for us. I would prefer not to have any default value for deadline.